fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/FoldConstantRule.java

```diff
@@ -0,0 +1,440 @@
+                whenClauses.add(new WhenClause(whenOperand, rewrite(whenClause.getResult(), context)));
+            } else if (BooleanLiteral.TRUE.equals(whenOperand)) {
+                foundNewDefault = true;
+                newDefault = rewrite(whenClause.getResult(), context);
+                break;
+            }
+        }
+
+        Expression defaultResult;
+        if (foundNewDefault) {
+            defaultResult = newDefault;
+        } else {
+            defaultResult = rewrite(caseWhen.getDefaultValue().orElse(Literal.of(null)), context);
+        }
+
+        if (whenClauses.isEmpty()) {
+            return defaultResult;
+        }
+        return new CaseWhen(whenClauses, defaultResult);
+    }
+
+    @Override
+    public Expression visitInPredicate(InPredicate inPredicate, ExpressionRewriteContext context) {
+        List<Expression> newChildren = inPredicate.children().stream().map(c -> rewrite(c, context))
+                .collect(Collectors.toList());
+        if (newChildren.get(0) instanceof NullLiteral) {
+            return Literal.of(null);
+        }
+        if (isAllLiteral(newChildren)) {
+            Literal c0 = (Literal) newChildren.get(0);
+            for (int i = 1; i < newChildren.size(); i++) {
+                if (c0.compareTo((Literal) newChildren.get(i)) == 0) {
+                    return Literal.of(true);
+                }
+            }
+            return Literal.of(false);
+        }
+        return inPredicate.withChildren(newChildren);
+    }
+
+    @Override
+    public Expression visitIsNull(IsNull isNull, ExpressionRewriteContext context) {
+        Expression child = rewrite(isNull.child(), context);
+        if (child instanceof NullLiteral) {
+            return Literal.of(true);
+        } else if (child instanceof Literal) {
+            return Literal.of(false);
+        }
+        return isNull.withChildren(child);
+    }
+
+    @Override
+    public Expression visitTimestampArithmetic(TimestampArithmetic arithmetic, ExpressionRewriteContext context) {
+        Expression left = rewrite(arithmetic.child(0), context);
+        Expression right = rewrite(arithmetic.child(1), context);
+        if (isAllLiteral(left, right)) {
+            return ExpressionEvaluator.INSTANCE.eval(arithmetic.withChildren(left, right));
+        }
+        return arithmetic.withChildren(left, right);
+    }
+
+    private Expression foldByBe(Expression root, ExpressionRewriteContext context) {
+        if (root.isConstant() && !root.isLiteral()) {
+            Expr expr = ExpressionTranslator.INSTANCE.translate(root, null);
+            assignId(expr);
+            Map<String, Expr> ori = new HashMap<>();
+            ori.put(expr.getId().toString(), expr);
+
+            Map<String, Map<String, TExpr>> paramMap = new HashMap<>();
+            Map<String, TExpr> constMap = new HashMap<>();
+
+            collectConst(expr, constMap);
+            paramMap.put("0", constMap);
```

Comment:
what is the "0" mean?

Reply:
This is just a token, so it doesn't make sense, but it does get confusing, so use the ID of the current expression instead